### ISSUE TITLE
🚀 Roadmap

### ISSUE THREAD

**Issue by roaldarbol** [roaldarbol/Bux#41]
Alright, I think Bux can still be of use. But I would want a completely refactor, make it easier to maintain and use modern components.

## Tool Summary
- **API**: [FastAPI](https://fastapi.tiangolo.com/)
- **CLI**: [Typer](https://typer.tiangolo.com/)
- **GUI**: QT6, PyQt6/PySide6?
- **Registry**: [autoregistry](https://github.com/BrianPugh/autoregistry)
- **Cameras**: 
	- **Webcams**: [OpenCV](https://github.com/orgs/opencv/projects/1), OpenCV 5 is under development.
	-  **Raspberry Pi CSI cameras**: [picamera2](https://github.com/raspberrypi/picamera2)
	- **GoPro**: [Open GoPro](https://gopro.github.io/OpenGoPro/python_sdk/)
	- **Wireless iPhone/Android cameras**: OpenCV
	- **ISP (surveillance) cameras**: OpenCV
	- **FLIR**: Spinnaker SDK
- **Microcontrollers**: 
	- **Micropython**: [Belay](https://github.com/BrianPugh/belay)
	- **Arduino**: pyFirmata
- **Packaging**: [pixi](https://pixi.sh) and [rattler-build](https://prefix-dev.github.io/rattler-build/latest/)
- (**Executable**: [Nuitka](https://nuitka.net/), but will stick to conda as long as I can)

## Microcontrollers
I'll need to come up with a proper solution for how to integrate with Belay. 

## Cameras
I think the camera solution worked well, but a few things to consider:
- More camera types should be supported long-term
- Functions should be made separately for each camera type (e.g. GoPro, FLIR, iPhone, RPi CSI cams). 
- Instead of OpenCV and Picamera2, [VidGear](https://github.com/abhiTronix/vidgear) could be considered. 
- Expose as many settings as possible

**Comment by BrianPugh** [comment]
shameless self-plug; checkout my project [Cyclopts](https://github.com/BrianPugh/cyclopts) for the CLI ;)

However, I've rarely actually used FastAPI; if Typer/FastAPI play nicer together (since they're both Tiangolo projects), then I'd say go for Typer.

Edit: also more self-plugs checkout my [autoregistry](https://github.com/BrianPugh/autoregistry) for managing your camera-hooks. An example (with Belay) can be seen in my [magnetometer](https://github.com/BrianPugh/magnetometer) project where I use autoregistry to manage my different magnetic-sensor-types.

**Comment by roaldarbol** [comment]
@BrianPugh Haha, great to see you here! You may have noticed I also just started a discussion on Belay 😂 I'm still not 100% sure I'll need FastAPI (it's been recommended to me, but I'm not really sure I see the need for myself) - so will definitely consider Cyclopts! `autoregistry` also really looks like something I could make use of! I'm assuming that oculd e.g. be good for having a registry of camera types/registered cameras?
Are they on conda-forge? I'm hoping to mostly rely on that rather than PyPi. I'm a big user of pixi, so can highly recommend that and rattler-build (until rattler-build becomes bundled into pixi).

**Comment by BrianPugh** [comment]
> I'm assuming that could e.g. be good for having a registry of camera types/registered cameras?

Precisely; assuming that each camera needs custom data/methods/whatever, this is exactly what `autoregistry` was made for.

All my projects are on pypi; I'm not against publishing on conda-forge, I just never really investigated it. If it becomes a big blocker I could maybe investigate more.

**Comment by roaldarbol** [comment]
That sounds great, I'll add it to the list! 

I could possibly help with that, come the time - I'm starting to get somewhat of a grip on how that works (I think...)😊

EDIT: Seems they're already on there: 
- https://prefix.dev/channels/conda-forge/packages/belay
- https://prefix.dev/channels/conda-forge/packages/autoregistry
- https://prefix.dev/channels/conda-forge/packages/cyclopts

So I guess you can add `mamba install` to your projects!

**Comment by roaldarbol** [comment]
Note to self, maybe it would be much simpler to make a napari plugin?! E.g. https://github.com/jacopoabramo/napari-live-recording already shows you can record, but is more focused on microscopy.
